pkg/genericapiserver/options/server_run_options.go

```diff
@@ -212,15 +213,26 @@ func mergeGroupVersionIntoMap(gvList string, dest map[string]unversioned.GroupVe
 }
 
 // Returns a clientset which can be used to talk to this apiserver.
-func (s *ServerRunOptions) NewSelfClient() (clientset.Interface, error) {
+func (s *ServerRunOptions) NewSelfClient() (clientset.Interface, string, error) {
 	clientConfig := &restclient.Config{
-		Host: net.JoinHostPort(s.InsecureBindAddress.String(), strconv.Itoa(s.InsecurePort)),
 		// Increase QPS limits. The client is currently passed to all admission plugins,
 		// and those can be throttled in case of higher load on apiserver - see #22340 and #22422
 		// for more details. Once #22422 is fixed, we may want to remove it.
 		QPS:   50,
 		Burst: 100,
 	}
+	if s.InsecurePort == 0 {
+		clientConfig.Host = "https://" + net.JoinHostPort(s.BindAddress.String(), strconv.Itoa(s.SecurePort))
+		clientConfig.Insecure = true
+		clientConfig.BearerToken = uuid.NewRandom().String()
+		clientConfig.TLSClientConfig = restclient.TLSClientConfig{
+			CertFile: s.TLSCertFile,
+			KeyFile:  s.TLSPrivateKeyFile,
```

Comment:
Anything that has a kubeconfig (kubelet, kube proxy, controller manager, etc) has it, but I don't think the server knows it's own CA bundle currently